File: Ch01/ex1_07.exs
# SICP in Elixir: Redefine tolerance to vary based on the guess,
# rather than being a constant value.

defmodule Ex1_07 do
  def sqrt(x), do: sqrt(1.0, x)

  # Private functions
  defp sqrt(guess, x) do
    case good_enough(guess, x) do
      true -> guess
      false -> sqrt(improve(guess, x), x)
    end
  end

  defp average(x, y) do
    (x + y) / 2.0
  end

  defp good_enough(guess, x) do
    abs(square(guess) - x) < tolerance(x)
  end

  defp improve(guess, x) do
    average(guess, x / guess)
  end

  defp square(x), do: x * x
  defp tolerance(x), do: x * 0.001
end

IO.puts(Ex1_07.sqrt(2))

File: Ch01/ex1_07.erl
-module(ex1_07).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Redefine tolerance to vary based on the guess, not constant value.").
-export([sqrt/1]).

sqrt(X) -> sqrt(1.0, X).

%% HELPER FUNCTIONS

sqrt(Guess, X) ->
    % calling a function within a guard expression seems to be
    % illegal, so we stick with a case here.
    case good_enough(Guess, X) of
        true  -> Guess;
        false -> sqrt(improve(Guess, X), X)
    end.

good_enough( Guess, X ) -> (abs(square(Guess) - X) < tolerance(X)).
improve(     Guess, X ) -> average(Guess, (X/Guess)).

average(X, Y) -> (X + Y) / 2.0.
square(X)     -> X * X.
tolerance(X)  -> X * 0.001.

